File: 005_reverse_binary_tree/reverse_binary_tree.rb
class Node
  attr_accessor :value, :left, :right

  def initialize(v, l=nil, r=nil)
    @value = v
    @left = l
    @right = r
  end

  def to_s
    "Value: #{@value}, Left: #{@left.value}, Right: #{@right.value}"
  end
end

def solution1(root)
  # Recursively mutates the original tree
  return unless root

  tmp = root.left
  root.left = root.right
  root.right = tmp

  solution1(root.left)
  solution1(root.right)
end

def solution2(root)
  # Recursively creates a new tree
  return nil unless root
  return Node.new(root.value, solution2(root.right), solution2(root.left))
end

if __FILE__ == $0
  tree = Node.new(1)
  tree.left = Node.new(2)
  tree.right = Node.new(3)
  solution1(tree)

  puts tree
  puts solution2(tree)
end

File: 005_reverse_binary_tree/reverse_binary_tree.py
class Node(object):
    def __init__(self, v, l=None, r=None):
        self.value = v
        self.left = l
        self.right = r

    def __str__(self):
        return 'Value: %s, Left: %s, Right: %s' % (
            self.value, self.left.value, self.right.value
        )

def solution1(root):
    """
    Recursively mutates the original tree
    """
    if not root:
        return

    tmp = root.left
    root.left = root.right
    root.right = tmp

    solution1(root.left)
    solution1(root.right)

def solution2(root):
    """
    Recursively creates a new tree
    """
    if not root:
        return None

    return Node(root.value, solution2(root.right), solution2(root.left))

if __name__ == '__main__':
    tree = Node(1)
    tree.left = Node(2)
    tree.right = Node(3)
    solution1(tree)

    print(tree)
    print(solution2(tree))
